### ISSUE TITLE
extracting widget doesn't create members

### ISSUE THREAD

**Issue by stephane-archer** [dart-lang/sdk#59819]
# Input:
``` Dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:lutme/adaptive_image.dart';
import 'package:lutme/exposure_analysis_tool_change_notifier.dart';
import 'package:lutme/exposure_analysis_widget.dart';
import 'package:lutme/img_with_before_and_after_click.dart';
import 'package:lutme/lut_intensity_selector.dart';
import 'package:lutme/preview_generator_change_notifier.dart';
import 'package:lutme/ref_image_transform_provider.dart';
import 'package:lutme/selected_lut_change_notifier.dart';

class MainPreview extends ConsumerWidget {
  const MainPreview({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    AsyncValue<String?> imagePathAsyncValue = ref.watch(refImageTransformProvider);
   return imagePathAsyncValue.when<Widget>(data: (imagePath) {
// START EXTRACTING WIDGET
      SelectedLutChangeNotifier selectedLutChangeNotifier =
          ref.watch(selectedLutProvider);
      ExposureAnalysisToolChangeNotifier exposureAnalysisToolChangeNotifier =
          ref.watch(exposureAnalysisToolProvider);
      PreviewGeneratorChangeNotifier previewGeneratorChangeNotifier =
          ref.watch(previewGeneratorProvider);
      if (imagePath == null) {
        return const SizedBox.shrink();
      }
      String? pathImgSelectedLut = selectedLutChangeNotifier.pathImgSelectedLut;
      if (pathImgSelectedLut == null) {
        if (exposureAnalysisToolChangeNotifier.isActivated) {
          return ExposureAnalysisWidget(
              imagePath, previewGeneratorChangeNotifier);
        }
        return AdaptiveImage(FileImage(File(imagePath)), false);
      }
      if (exposureAnalysisToolChangeNotifier.isActivated) {
        return ExposureAnalysisWidget(
            pathImgSelectedLut, previewGeneratorChangeNotifier);
      }
      return Column(
        children: [
          Expanded(
            flex: 8,
            child: ImgWithBeforeAndAfterClick(
                before: AdaptiveImage(FileImage(File(imagePath)), false),
                after: OpacityStack(
                  before: AdaptiveImage(FileImage(File(imagePath)), false),
                  after:
                      AdaptiveImage(FileImage(File(pathImgSelectedLut)), false),
                )),
          ),
          const SizedBox(height: 10.0),
          const LutIntensitySelector(),
        ],
      );
// END EXTRACTING WIDGET
    }, error: (e, s) {
      return Text("Error: $e");
    }, loading: () {
      return const CircularProgressIndicator();
    });
  }
}

```
# output:
``` Dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:lutme/adaptive_image.dart';
import 'package:lutme/exposure_analysis_tool_change_notifier.dart';
import 'package:lutme/exposure_analysis_widget.dart';
import 'package:lutme/img_with_before_and_after_click.dart';
import 'package:lutme/lut_intensity_selector.dart';
import 'package:lutme/preview_generator_change_notifier.dart';
import 'package:lutme/ref_image_transform_provider.dart';
import 'package:lutme/selected_lut_change_notifier.dart';

class MainPreview extends ConsumerWidget {
  const MainPreview({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    AsyncValue<String?> imagePathAsyncValue = ref.watch(refImageTransformProvider);
   return imagePathAsyncValue.when<Widget>(data: (imagePath) {
      return _MainPreviewOnData();
    }, error: (e, s) {
      return Text("Error: $e");
    }, loading: () {
      return const CircularProgressIndicator();
    });
  }
}

class _MainPreviewOnData extends StatelessWidget {
  const _MainPreviewOnData({
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    SelectedLutChangeNotifier selectedLutChangeNotifier =
        ref.watch(selectedLutProvider);
    ExposureAnalysisToolChangeNotifier exposureAnalysisToolChangeNotifier =
        ref.watch(exposureAnalysisToolProvider);
    PreviewGeneratorChangeNotifier previewGeneratorChangeNotifier =
        ref.watch(previewGeneratorProvider);
    if (imagePath == null) {
      return const SizedBox.shrink();
    }
    String? pathImgSelectedLut = selectedLutChangeNotifier.pathImgSelectedLut;
    if (pathImgSelectedLut == null) {
      if (exposureAnalysisToolChangeNotifier.isActivated) {
        return ExposureAnalysisWidget(
            imagePath, previewGeneratorChangeNotifier);
      }
      return AdaptiveImage(FileImage(File(imagePath)), false);
    }
    if (exposureAnalysisToolChangeNotifier.isActivated) {
      return ExposureAnalysisWidget(
          pathImgSelectedLut, previewGeneratorChangeNotifier);
    }
    return Column(
      children: [
        Expanded(
          flex: 8,
          child: ImgWithBeforeAndAfterClick(
              before: AdaptiveImage(FileImage(File(imagePath)), false),
              after: OpacityStack(
                before: AdaptiveImage(FileImage(File(imagePath)), false),
                after:
                    AdaptiveImage(FileImage(File(pathImgSelectedLut)), false),
              )),
        ),
        const SizedBox(height: 10.0),
        const LutIntensitySelector(),
      ],
    );
  }
}

```
`ref` and `imagePath` doesn't exist
# expected code:
``` Dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:lutme/adaptive_image.dart';
import 'package:lutme/exposure_analysis_tool_change_notifier.dart';
import 'package:lutme/exposure_analysis_widget.dart';
import 'package:lutme/img_with_before_and_after_click.dart';
import 'package:lutme/lut_intensity_selector.dart';
import 'package:lutme/preview_generator_change_notifier.dart';
import 'package:lutme/ref_image_transform_provider.dart';
import 'package:lutme/selected_lut_change_notifier.dart';

class MainPreview extends ConsumerWidget {
  const MainPreview({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    AsyncValue<String?> imagePathAsyncValue =
        ref.watch(refImageTransformProvider);
    return imagePathAsyncValue.when<Widget>(data: (imagePath) {
      return _MainPreviewOnData(imagePath);
    }, error: (e, s) {
      return Text("Error: $e");
    }, loading: () {
      return const CircularProgressIndicator();
    });
  }
}

class _MainPreviewOnData extends ConsumerWidget {
  final String? _imagePath;
  const _MainPreviewOnData(
    this._imagePath,
  );

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    SelectedLutChangeNotifier selectedLutChangeNotifier =
        ref.watch(selectedLutProvider);
    ExposureAnalysisToolChangeNotifier exposureAnalysisToolChangeNotifier =
        ref.watch(exposureAnalysisToolProvider);
    PreviewGeneratorChangeNotifier previewGeneratorChangeNotifier =
        ref.watch(previewGeneratorProvider);
    String? imagePath = _imagePath;
    if (imagePath == null) {
      return const SizedBox.shrink();
    }
    String? pathImgSelectedLut = selectedLutChangeNotifier.pathImgSelectedLut;
    if (pathImgSelectedLut == null) {
      if (exposureAnalysisToolChangeNotifier.isActivated) {
        return ExposureAnalysisWidget(
            imagePath, previewGeneratorChangeNotifier);
      }
      return AdaptiveImage(FileImage(File(imagePath)), false);
    }
    if (exposureAnalysisToolChangeNotifier.isActivated) {
      return ExposureAnalysisWidget(
          pathImgSelectedLut, previewGeneratorChangeNotifier);
    }
    return Column(
      children: [
        Expanded(
          flex: 8,
          child: ImgWithBeforeAndAfterClick(
              before: AdaptiveImage(FileImage(File(imagePath)), false),
              after: OpacityStack(
                before: AdaptiveImage(FileImage(File(imagePath)), false),
                after:
                    AdaptiveImage(FileImage(File(pathImgSelectedLut)), false),
              )),
        ),
        const SizedBox(height: 10.0),
        const LutIntensitySelector(),
      ],
    );
  }
}

```

**Comment by FMorschel** [comment]
<details>

<summary> Personal request to author </summary>

@stephane-archer, this is what I meant when I asked you in another issue to simplify your examples. Although this refactor (extract widget) has only a few reports (many of them opened by you, thanks, by the way!) and I could manually look through them all, this one is exactly the same problem as https://github.com/dart-lang/sdk/issues/61546.

I tend to be thorough and look at other similar reports on the issue tracker, but even I miss them sometimes. If you had spent some time looking at what exactly was happening here, shortening your repro, you could have gotten to a repro like I got at https://github.com/dart-lang/sdk/issues/61546#issuecomment-3351695959, and you would not have needed to open yet another issue (although it helped that I now saw it and had the time to fix it, it usually stays open for a while still).

I don't mean to be rude, that is not the message I'm trying to send, I'm merely asking you, if you have the time (even as a new comment) to simplify your repro (even more so if you can use only default classes, not package ones), it would help us _a lot_. This way, we can spend more time fixing other reports (even yours), and not have to verify for ourselves how to simplify your code (to find the cause of the issue), and if this report is actually the same problem (because they are vague) as another one.

</details>

Thanks again for reporting. We really appreciate it!

Here is the CL for this https://dart-review.googlesource.com/c/sdk/+/452483

**Comment by stephane-archer** [comment]
@FMorschel Thank you for the clarification and for all the fixes you have been working on. The refactoring experience already feels much better than when I first started reporting these issues.

I do understand your request about simplifying repros and I will try to do that when I can. The reality is that I usually run into these bugs in the middle of my work and my main priority is getting my tasks done. Since I am not paid to file bug reports I do not always have the time to reduce the problem into a perfectly minimal example on the spot.

That said I will do my best to provide simpler repros whenever possible and I really appreciate the time you and the team spend investigating these issues and making Dart better for everyone. Thank you again.

Should I close this issue?

**Comment by FMorschel** [comment]
> I do understand your request about simplifying repros and I will try to do that when I can. The reality is that I usually run into these bugs in the middle of my work and my main priority is getting my tasks done. Since I am not paid to file bug reports I do not always have the time to reduce the problem into a perfectly minimal example on the spot.
>
> That said I will do my best to provide simpler repros whenever possible [...]

That is all I could ask for, thank you for understanding. Even opening these already helps a lot (see what I mean about later - look at the time difference between the opening and comment - coming up with a smaller repro - this is not even "smaller", it is _the_ repro https://github.com/dart-lang/sdk/issues/61382), I merely ask you to help us as much as possible. We do have a large number of issues to handle, so any help is really appreciated!

> Should I close this issue?

Again, the CL description does mention this issue and will close this once it lands. Feel free to do that if you want to, I just ask that you either leave this or https://github.com/dart-lang/sdk/issues/61546 open so the CL can notify us when that actually landed, so other people find these instead of filing new issues.

That said, if you want to rename either to make it clearer what the issue is about, it might be good. Something like: "Extract widget doesn't handle parameters" or whatever you prefer.
